File: PRACTICAS/CAMBIAR_RUBY/RUBY2/Civitas/lib/casilla.rb
require './tipo_sorpresa'
require './tipo_casilla'
require './sorpresa'
module Civitas
  class Casilla
    
    @@carcel = 8
    
    attr_accessor :nombre, :tipo, :importe, :titulo_propiedad, :mazo, :sorpresa
    def initialize (nombre = "sin_nombre", carcel = 8, importe = 0, tipo = nil, titulo_propiedad = TituloPropiedad.new("Ninguna", 0, 0, 0, 0, 0) , mazo = nil, sorpresa = nil)
      
      @nombre = nombre
      @tipo = tipo
      @importe = importe
      @@carcel = carcel
      @titulo_propiedad = titulo_propiedad
      @mazo = mazo
      @sorpresa = sorpresa
      
    end
    
    def self.casilla(nombre)
    self.new(nombre, 8, 0, TipoCasilla::DESCANSO, TituloPropiedad.new("Ninguna", 0, 0, 0, 0, 0), nil, nil)
  end

  def self.casilla2(titulo_propiedad )

    self.new(titulo_propiedad.nombre, 8, titulo_propiedad.precio_compra, TipoCasilla::CALLE, titulo_propiedad, nil, nil)

  end

  def self.casilla3(cantidad,nombre)

    self.new(nombre, 8, cantidad, TipoCasilla::IMPUESTO, TituloPropiedad.new("Ninguna", 0, 0, 0, 0, 0), nil, nil)

  end

  def self.casilla4(num_casilla_carcel, nombre)

    self.new(nombre, num_casilla_carcel, 0, TipoCasilla::JUEZ, TituloPropiedad.new("Ninguna", 0, 0, 0, 0, 0), nil, nil)

  end

  def self.casilla5(mazo, nombre)

    self.new(nombre, 8, 0, TipoCasilla::SORPRESA, TituloPropiedad.new("Ninguna", 0, 0, 0, 0, 0), mazo, nil)

  end

  
  private
    def informe(actual, todos)
      if(TipoCasilla::CALLE == @tipo)
        Diario.instance.ocurre_evento(todos[actual].nombre + " ha caido en la casilla: #{to_s} que tiene la propiedad: #{@titulo_propiedad.to_s}")
      else
        Diario.instance.ocurre_evento(todos[actual].nombre + " ha caido en la casilla: #{to_s} ")
      end
    end
  
    public
    def to_s

    return "#{@nombre}, la carcel esta en la posicion #{@@carcel} y es de tipo #{@tipo}\n"

    end
    
    
    def jugador_correcto(iactual,todos)

      return (iactual >= 0 && iactual < todos.size)

    end
    public
    def recibe_jugador(iactual,todos)

              case @tipo
                when TipoCasilla::CALLE
                  recibe_jugador_calle(iactual, todos)

               when TipoCasilla::IMPUESTO
                  recibe_jugador_impuesto(iactual, todos)

              when TipoCasilla::JUEZ
                  recibe_jugador_juez(iactual, todos)

              when TipoCasilla::SORPRESA
                  recibe_jugador_sorpresa(iactual, todos)
              else
                informe(iactual,todos) 
              end

    end
    
    private
  def recibe_jugador_calle(iactual,todos)

        if jugador_correcto(iactual, todos)
            informe(iactual, todos)

            if !@titulo_propiedad.tiene_propietario

              todos.at(iactual).puede_comprar_casilla
            else

                @titulo_propiedad.tramitar_alquiler(todos.at(iactual))

            end
        end
  end


  def recibe_jugador_impuesto(iactual,todos)

    if (jugador_correcto(iactual, todos))
      informe(iactual, todos)
      todos.at(iactual).paga_impuesto(@importe)
    end

  end

  def recibe_jugador_juez(iactual,todos)

    if (jugador_correcto(iactual, todos))
      informe(iactual, todos)
      todos.at(iactual).encarcelar(@@carcel)
    end
  end
    
    def recibe_jugador_sorpresa(iactual,todos)

      if jugador_correcto(iactual,todos)

              @sorpresa = @mazo.siguiente
              informe(iactual,todos)
              @sorpresa.aplicar_a_jugador(iactual, todos)

      end

    end
    
  end
  end

File: PRACTICAS/practica1/Civitas_P1/src/civitas_p1/Casilla.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;

import java.util.ArrayList;

/**
 *
 * @author angelsolano
 */
public class Casilla {
    
    private String nombre;
    private TipoCasilla tipo;
    private static int carcel;
    private float importe;
    private TituloPropiedad tituloPropiedad; //solo para CALLE
    private MazoSorpresas mazo;
    private Sorpresa sorpresa;
    
    Casilla(String nombre , TipoCasilla tipo){
        init();
        this.nombre=nombre;
        this.tipo=tipo;
    }
    Casilla(TituloPropiedad titulo){
        init();
        tituloPropiedad=titulo;
    }
    Casilla(String name){
        init();
        this.nombre = name;
    }
    Casilla(float cantidad, String nombre){
        init();
        this.nombre = nombre;
        importe=cantidad;
    }
    Casilla(int numCasillaCarcel, String nombre){
        init();
        this.nombre=nombre;
        carcel=numCasillaCarcel;
    }
    Casilla(MazoSorpresas mazo,String nombre){
        init();
        this.mazo=mazo;
        this.nombre=nombre;
        
    }
    private void init(){
        
        nombre = "Calle";
        tipo = TipoCasilla.CALLE;
        carcel = 0;
        importe = (float) 0.0;
        TituloPropiedad tl;
        tl = new TituloPropiedad("Propiedad vacia",(float)0.0,0,(float)0.0,(float)0.0,(float)0.0);
        tituloPropiedad = tl ;
        MazoSorpresas maz = null;
        maz.init();
        mazo = maz;
        Sorpresa sor;
        sor = new Sorpresa(TipoSorpresa.IRCASILLA,mazo);
        sorpresa = sor;
        
    }
    private void informe(int actual, ArrayList<Jugador>todos){
        
        Diario.getInstance().ocurreEvento("El jugador que ha caido en la casilla es " + todos.get(actual).getNombre());
        toString();
        
    }
    public boolean jugadorCorrecto(int actual, ArrayList<Jugador>todos){
        
        if(actual>=0 && actual<todos.size()){
            
            return true;
            
        }
        
        return false;
    }
    private void recibeJugador_impuesto(int actual,ArrayList<Jugador>todos){
        
        if(jugadorCorrecto(actual,todos)){
            
            informe(actual,todos);
            todos.get(actual).pagaImpuesto(importe);
            
        }
        
    }
    private void recibeJugador_juez(int actual, ArrayList<Jugador>todos){
        
        if(jugadorCorrecto(actual,todos)){
            
            informe(actual,todos);
            todos.get(actual).encarcelar(carcel);
            
        }
        
    }
    public String toString(){
        
        return "Esta casilla se llama" +nombre+ ", es de tipo" + tipo + ", la carcel se encuentra en la posicion " +carcel+ ", el importe a pagar es de " +importe;
        
    }
    public String getNombre(){
        return this.nombre;
    }
    
    TituloPropiedad getTituloPropiedad(){
        
        return tituloPropiedad;
        
    }
    
}
